README
HTML-Tree version 5.909, released May 24, 2014
THIS IS A DEVELOPMENT RELEASE FOR TESTING PURPOSES ONLY.


This distribution contains a suite of modules for representing,
creating, and extracting information from HTML syntax trees; there is
also relevent documentation.  These modules used to be part of the
libwww-perl distribution, but are now unbundled in order to facilitate
a separate development track.  Bug reports and discussions about these
modules can be sent to the RT queue at <bug-html-tree@rt.cpan.org>.

The modules and documentation present in this collection are:

  HTML::Tree -- overview of the modules.

  HTML::Tree::AboutObjects -- an article by Sean M. Burke from
    The Perl Journal #17: "User's View of Object-Oriented Modules"

  HTML::Tree::AboutTrees -- an article by Sean M. Burke from
    The Perl Journal #18: "Trees", about tree-shaped data
    structures in Perl.

  HTML::Tree::Scanning -- an article by Sean M. Burke from
    The Perl Journal #19: "Scanning HTML".


  HTML::Element - class for representing the nodes of the HTML syntax
        trees.  The elements have other elements and text segments
        as children.  The HTML::Element class has methods to build,
        alter, and traverse the structure of the tree.

  HTML::TreeBuilder - a class that uses HTML::Parser to read HTML
        document text and build from it a syntax tree made of
        HTML::Element nodes.

  HTML::Parse - deprecated module.  Now just a wrapper around
        HTML::TreeBuilder

  HTML::AsSubs - module providing an easy way to build an HTML syntax
        tree by nesting functions.



INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install



DEPENDENCIES

  Package         Minimum Version
  --------------- ---------------
  perl             5.8.0
  Carp             
  Encode           
  Exporter         
  HTML::Entities   
  HTML::Parser     3.46
  HTML::Tagset     3.02
  IO::HTML         
  Scalar::Util



SOURCE

The source is now kept at https://github.com/madsen/HTML-Tree



SUBLIMINAL MESSAGE: BUY THE BOOK!

The book /Perl & LWP/ by Sean M. Burke is published by
O'Reilly and Associates, 2002.  ISBN: 0-596-00178-9

It has several chapters to do with HTML processing in general,
and HTML-Tree specifically.  There's more info at:

  http://www.oreilly.com/catalog/perllwp/
  http://www.amazon.com/exec/obidos/ASIN/0596001789



COPYRIGHT

Copyright 1995-1998 Gisle Aas; 1999-2004 Sean M. Burke;
2005 Andy Lester; 2006 Pete Krawczyk; 2010 Jeff Fearn;
2012 Christopher J. Madsen (Except the articles contained in
HTML::Tree::AboutObjects, HTML::Tree::AboutTrees, and
HTML::Tree::Scanning, which are all copyright 2000 The Perl Journal.)

Except for those three TPJ articles, the whole HTML-Tree distribution,
of which this file is a part, is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

Those three TPJ articles may be distributed under the same terms as
Perl itself.

The programs in this library are distributed in the hope that they
will be useful, but without any warranty; without even the implied
warranty of merchantability or fitness for a particular purpose.


Changes
Changelog for HTML-Tree

5.909    2014-05-24  Trial Release by Christopher J. Madsen

    [TESTS]
    - Skip LWP::UserAgent tests if the version installed is too old
    - Change 00system.t to use diag so the output is visible

5.908    2014-05-17  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - New as_lol & content_as_lol methods in HTML::Element
      (Ben Deutsch, GitHub #2)

    [FIXES]
    - as_XML inserted extra whitespace in comments (Tomaž Solc, RT #94311)

5.907    2013-09-13  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - New content_as_HTML & content_as_XML methods in HTML::Element

5.906    2013-07-13  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - Explain that implicit_tags => 0 still returns implicit <html> (RT #86062)
    - typo fixes supplied by David Steinbrunner (GitHub #1)

    [TESTS]
    - Add implicit-tags-off.t

5.905    2013-06-29  Trial Release by Christopher J. Madsen

    [FIXES]
    - as_XML would not allow single-character attribute names (RT #86566)

5.904    2013-06-08  Trial Release by Christopher J. Madsen

    [FIXES]
    - as_text would infinite loop if it encountered undef (RT #83757)

5.903    2013-06-01  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - $HTML::Element::default_encoding gets its default value from
      $ENV{PERL_HTML_TREE_ENCODING}, so you can restore the old behavior
      without modifying any code (just export PERL_HTML_TREE_ENCODING='').

5.902    2013-03-01  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - minor clarifications

    [TESTS]
    - Add dump.t

5.901    2012-12-22  Trial Release by Christopher J. Madsen

    [TESTS]
    - Fix encoding.t to work on Perl 5.8.9 and 5.10.0

5.900    2012-12-15  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - parse_file (and new_from_file) now try to determine the encoding
      automatically when given a filename (not a filehandle).  To restore
      the old behavior, set the new encoding attribute to the empty string.
      To restore it globally, set $HTML::Element::default_encoding = ''.

    [ENHANCEMENTS]
    - new_from_file & new_from_url now let you set parsing attributes
    - New shortcut constructor new_from_string is like new_from_content,
      but allows you to set parsing attributes
    - New shortcut constructor new_from_http for constructing a tree
      from the content of a HTTP::Message (or subclass like HTTP::Response)
    - Setting the new self_closed_tags attribute to 1 makes TreeBuilder
      handle XML-style self-closed tags (e.g. <a id="a1" />)
    - New child_nodes method makes for simpler recursion
    - New openw and encode_fh methods for writing a file with the
      correct encoding

    [DOCUMENTATION]
    - new actually does take optional attributes
      (It has since at least 3.18, although undocumented, and it did
       not previously work with ignore_ignorable_whitespace.)
    - methods & attributes added in version 4.0 or later are now marked
    - don't recommend the traverse method; give recursive example (RT #48344)

    [TESTS]
    - Add test for self_closed_tags attribute.
    - Clarify skip message in construct_tree.t (RT #81371)


5.03     2012-09-22  Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - as_HTML no longer indents <textarea> (Tomohiro Hosaka) (RT #70385)

    [FIXES]
    - as_trimmed_text did not accept '0' for extra_chars

    [DOCUMENTATION]
    - Explain that as_text never adds whitespace (RT #66498)
    - Explain what extra_chars can contain for as_trimmed_text.


5.02     2012-06-27  Release by Christopher J. Madsen

    [TESTS]
    - Do not attempt to check result of $! in construct_tree.t
      (The fix in 5.01 was not successful.)


5.01     2012-06-20  Release by Christopher J. Madsen

    [TESTS]
    - Force C locale in construct_tree.t (in non-English locales,
      $! will produce messages in a different language) (RT #77823)
    - Add test for preserving whitespace while parsing.


5.00     2012-06-12  Release by Christopher J. Madsen

   There are only some minor documentation changes since 4.903.
   This is a summary of the most significant changes since 4.2.

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - Use weak references to avoid memory leaks
      See "Weak References" in HTML::Element for details.
    - new_from_file now dies if the file cannot be opened.  $! records
      the specific problem.  (Previously, you got a tree with a few
      implicit elements.)
    - Some methods normally returning a scalar could return the empty
      list in certain circumstances.  This has been corrected.  The
      affected methods are: address, deobjectify_text, detach, is_inside,
      & pindex.
    - deprecate the Version sub/method.  Use the VERSION method instead.

    [ENHANCEMENTS]
    - add new_from_url (Using LWP) (David Pottage) (RT #68097)

    [DOCUMENTATION]
    - Explain that parse_file (and new_from_file) opens files in binary mode
      BUT THIS IS PLANNED TO CHANGE.


4.903    2012-06-08  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - Explain that parse_file (and new_from_file) opens files in binary mode
      BUT THIS IS PLANNED TO CHANGE.

    [TESTS]
    - test error handling for new_from_file & new_from_url
    - remove use_ok from most tests
      (if the module won't load, the tests can't pass anyway)


4.902    2012-06-06  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - new_from_url now dies if the request fails or the response is not HTML


4.901    2012-06-06  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - new_from_file now dies if the file cannot be opened.  $! records
      the specific problem.  (Previously, you got a tree with a few
      implicit elements.)

Build.PL
use 5.008;
use strict;
use warnings;
use inc::My_Build;

my $builder = inc::My_Build->new(
    module_name        => 'HTML::Tree',
    license            => 'perl',
    dist_author        => 'Christopher J. Madsen <perl@cjmweb.net>',
    dist_version_from  => 'lib/HTML/Element.pm',
  'build_requires' => {
    'Module::Build' => '0.2808',
    'Test::Fatal' => '0',
    'Test::More' => '0.88',
    'base' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.2808'
  },
  'recommends' => {
    'HTML::FormatText' => '0',
    'LWP::UserAgent' => '5.815'
  },
  'requires' => {
    'Carp' => '0',
    'Encode' => '0',
    'Exporter' => '0',
    'HTML::Entities' => '0',
    'HTML::Parser' => '3.46',
    'HTML::Tagset' => '3.02',
    'IO::HTML' => '0',
    'Scalar::Util' => '0',
    'perl' => '5.008'
  },
    add_to_cleanup =>
        [ 'HTML-Tree-*', 'tmp', 'blib', 'lwp-test-*', 'MANIFEST.bak' ],
    script_files => 'bin',
);

$builder->create_build_script();
